crates/orderbook/src/order_quoting.rs

```diff
@@ -428,18 +436,42 @@ pub enum FindQuoteError {
     #[error("quote not found")]
     NotFound(Option<QuoteId>),
 
+    #[error("quote does not match parameters")]
+    ParameterMismatch(QuoteData),
+
     #[error(transparent)]
     Other(#[from] anyhow::Error),
 }
 
 /// Fields for searching stored quotes.
+#[derive(Clone, Debug, Default, PartialEq)]
 pub struct QuoteSearchParameters {
     pub sell_token: H160,
     pub buy_token: H160,
     pub sell_amount: U256,
     pub buy_amount: U256,
     pub fee_amount: U256,
     pub kind: OrderKind,
+    pub from: H160,
+    pub app_data: AppId,
+}
+
+impl QuoteSearchParameters {
+    /// Returns true if the search parameter instance matches the specified
+    /// quote data.
+    fn matches(&self, data: &QuoteData) -> bool {
+        let amounts_match = match self.kind {
+            OrderKind::Buy => self.buy_amount == data.quoted_buy_amount,
+            OrderKind::Sell => {
+                self.sell_amount == data.quoted_sell_amount
+                    || self.sell_amount + self.fee_amount == data.quoted_sell_amount
```

Comment:
It's not clear to me why the `sell_amount` is allowed to be 2 different values.
Even if they could be equivalent wouldn't we want to enforce that `QuoteData` matches perfectly?

Reply:
This is because quoted amounts (so `QuoteData::quoted_{sell,buy}_amount`) can be one of two things:

1. If the quote is:
   ```
   { ... sell_amount_before_fee: $VALUE ... }
   ```
   Then we will compute a price estimate between `sell_amount_before_fee = sell_amount + fee_amount` in order get get a gas estimate, use that to compute a `fee_amount` and then a `sell_amount = sell_amount_before_fee - fee_amount` and then scale down the `quoted_{sell,buy}_amount`s at the same price. This is an approximation to avoid having to use multiple price requests for different amounts (we could re-compute the price for `sell_amount` after we have an initial `fee_amount` for example). In this case, we have `quoted_sell_amount = sell_amount + fee_amount`.
2. If the quote is:
   ```
   { ... sell_amount_after_fee: $VALUE ... }
   ```
   Then we will compute a price estimate for `sell_amount_before_fee = sell_amount` and just compute the `fee_amount` from the gas estimate of this trade. In this case, we want to sell exactly some `sell_amount` regardless of the fee, so we don't need to scale the `quoted_{sell,buy}_amount`s at all. In this case we have `quoted_sell_amount = sell_amount`.

So the `quoted_sell_amount` in the database can always be one of two values:
- `sell_amount + fee_amount` for `sell_amount_before_fee` quotes
- `sell_amount` for `sell_amount_after_fee` quotes

This is also why in the original `MinFeeStoring` implementation, we had `find_measurement_exact` for finding a fee measurement based on an exact amount, and `find_measurement_including_larger_amount` for finding the lowest fee measurement with an amount greater than or equal to the specified amount.

Reply:
Hopefully this will become more clear in the code once I implement @vkgnosis's suggestion https://github.com/cowprotocol/services/pull/288#discussion_r902281716

Reply:
I think I get it now. 👍